node/pkg/watchers/sui/watcher.go

```diff
@@ -503,25 +450,154 @@ func (e *Watcher) Run(ctx context.Context) error {
 
 	select {
 	case <-ctx.Done():
-		_ = ws.Close(websocket.StatusNormalClosure, "")
 		return ctx.Err()
 	case err := <-errC:
-		_ = ws.Close(websocket.StatusInternalError, err.Error())
 		return err
 	}
 }
 
-// fixSuiWsURL ensures the websocket scheme is properly specified
-func (e *Watcher) fixSuiWsURL(logger *zap.Logger) error {
-	u, _ := url.Parse(e.suiWS)
+func (w *Watcher) getEvents() (SuiEventResponseData, error) {
+	// Only get events newer than the last processed height
+	var retVal SuiEventResponseData
+	var nextCursor struct {
+		TxDigest string
+		EventSeq string
+	}
+	firstTime := true
+	done := false
+	for !done {
+		var reader io.Reader
+		if firstTime {
+			reader = strings.NewReader(
+				fmt.Sprintf(`{"jsonrpc":"2.0", "id": 1, "method": "suix_queryEvents", "params": [{ "MoveEventType": "%s" }, null, %d, %t]}`,
+					w.suiMoveEventType, w.maximumBatchSize, w.descendingOrder))
+		} else {
+			reader = strings.NewReader(
+				fmt.Sprintf(`{"jsonrpc":"2.0", "id": 1, "method": "suix_queryEvents", "params": [{ "MoveEventType": "%s" }, { "txDigest": "%s", "eventSeq": "%s" }, %d, %t]}`,
+					w.suiMoveEventType, nextCursor.TxDigest, nextCursor.EventSeq, w.maximumBatchSize, w.descendingOrder))
+		}
+		resp, err := http.Post(w.suiRPC, "application/json", reader) //nolint:noctx // TODO FIXME we should propagate context with Deadline here.
+		if err != nil {
+			return retVal, fmt.Errorf("suix_queryEvents failed to post: %w", err)
+		}
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return retVal, fmt.Errorf("suix_queryEvents failed to read: %w", err)
+		}
+		resp.Body.Close()
+
+		var res SuiEventResponse
+		err = json.Unmarshal(body, &res)
+		if err != nil {
+			return retVal, fmt.Errorf("suix_queryEvents failed to unmarshal body: %s, error: %w", string(body), err)
+		}
+		fmt.Println("Number of events:", len(res.Result.Data))
+		for _, datum := range res.Result.Data {
+			// Check if the event is newer than the last processed height
+			height, hErr := w.getCheckpointForEvent(datum)
```

Comment:
In the dashboard I only fetched the last event's checkpoint instead of each one, which will save you some RPC calls. You can then, later use `sui_multiGetTransactionBlocks` to get all of the checkpoints instead of making an RPC call for each one later on. i.e. just calculate `done` and append the results first, then after having fetched at least to the desired checkpoint (after this for loop), use that batch call to get all of the blocks at once.